packages/server/src/scripts/parse_and_cache_data.js

```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env node
+/* eslint-disable */
+
+require("dotenv").config();
+
+const { loadRecordsForUpload } = require('../arpa_reporter/services/records')
+const _ = require("lodash");
+
+const inquirer = require("inquirer");
+
+// This is so named to keep from accidentally typing "knex" out of habit and thus having some
+// queries inadvertently outside the transaction.
+const knexWithoutTransaction = require("../db/connection");
+const { getUpload, cacheRecords } = require("../arpa_reporter/db/uploads");
+
+async function main() {
+    if (!process.env.POSTGRES_URL) {
+        console.error("must specify POSTGRES_URL env variable");
+        return;
+    }
+    console.log(
+        "Parsing and caching upload data with postgres db: ",
+        process.env.POSTGRES_URL
+    );
+
+    const { uploadId } = await inquirer.prompt([
+        {
+            type: "input",
+            name: "uploadId",
+            message: "What is the uuid of the upload you want to parse and cache:",
+        },
+    ])
+
+    console.log("Parsing and caching upload with id:", uploadId);
+
+
+    const upload = await getUpload(uploadId)
+    const records = await loadRecordsForUpload(upload)
+
+    await cacheRecords(upload, records)
```

Comment:
I do see some advantage to caching the records like this, but I feel like it would keep the system a good deal simpler if we just cache the whole workbook object.

That is to say, the expensive part of the audit report operation is this one line:

https://github.com/usdigitalresponse/usdr-gost/blob/_staging/packages/server/src/arpa_reporter/services/records.js#L66-L70

If we abstract just that operation with an attempted cache read and fallback I don't think you need to tweak other parts of the record processors to accommodate running outside of the server context.

Reply:
I'd be open to exploring that, but I have a couple major questions first:

1. Are we really certain that the other XLSX methods (including 4 separate calls to `XLSX.utils.sheet_to_json` are not also slow and expensive?
2. Since the fact we are calling `XLSX.utils.sheet_to_json` implies that the sheet isn't yet in a valid json form, then we presumably can't use the jsonb data type in the database column. What type would you prefer for this? Just a plain serialized string of the object?

I'm also worried that storing the full sheet will bog down our json data with a bunch of unnecessary formatting info and tons of data about hidden sheets.

Reply:
Good clarifying questions!

1. I can validate this, but I'm pretty confident that's what my profiling exposed.  XLSX.read does some gnarly, repeteitive, on-the-fly regex compilation to actually consume the Microsoft .xlsx file format.  Once it's been transformed to the XLSX (library) workbook format, it is JavaScript data and can be manipulated and transformed much more cheaply.
2. `XLSX.utils.sheet_to_json` is a poorly named function.  From the docs, this method "converts a worksheet object to an array of JSON objects."  The workbook object is itself absolutely valid JSON already.  I have a demo (and screenshot) of one of these objects in #842!

Your last point is definitely a big advantage to caching the parsed records.

I like the idea of keeping the our record logic as ignorant of the cache's existence as possible.  And I also like the idea of only having to invalidate the cache when we update the parameters passed to XLSX.read itself.   But those advantages definitely come at the cost of a somewhat larger cache volume.  A sample workbook file that I saved to disk as JSON was about 1 MB in size.  I haven't measured the equivalent records yet, but I'm sure they are a good deal smaller!